### PR TITLE
Add annotated ion serializer support

### PR DESCRIPTION
**Pull request by simonz-bq**

### CODE REVIEW

**Review by byronlin13 — APPROVED** [review]

File: Amazon.Ion.ObjectMapper/IonSerializer.cs [hunk 1]
```diff
@@ -210,6 +212,23 @@ public void Serialize<T>(IIonWriter writer, T item)
             }
 
             Type type = item.GetType();
+
+            if (options.AnnotatedIonSerializers != null)
```
Comment (justing-bq) [inline Amazon.Ion.ObjectMapper/IonSerializer.cs:216]: Do these `AnnotatedIonSerializers` happen with primitive types or just objects?
If the latter then this logic should happen within the object logic.

File: Amazon.Ion.ObjectMapper.Test/IonSerializerTest.cs [hunk 2]
```diff
@@ -35,5 +36,16 @@ public void SerializesAndDeserializesLists()
...
+        [TestMethod]
+        public void AnnotatedIonSerializer()
+        {
+            var annotatedIonSerializer = new Dictionary<string, dynamic>();
+            annotatedIonSerializer.Add("Supra", new SupraSerializer());
+            var serializer = new IonSerializer(new IonSerializationOptions { AnnotatedIonSerializers = annotatedIonSerializer });
+            var stream = serializer.Serialize(TestObjects.a90);
+            var output = serializer.Deserialize<string>(stream);
+            Assert.AreEqual(TestObjects.a90.ToString(), output);
+        }
```
Comment (HomayounA) [inline Amazon.Ion.ObjectMapper.Test/IonSerializerTest.cs:49]: Add more tests to cover multiple serializers and edge cases.

File: Amazon.Ion.ObjectMapper/IonSerializer.cs [hunk 3]
```diff
@@ -243,6 +262,17 @@ public object Deserialize(IIonReader reader, Type type)
 
         public object Deserialize(IIonReader reader, Type type, IonType ionType)
         {
+            if (options.AnnotatedIonSerializers != null)
+            {
+                foreach (var valuePair in options.AnnotatedIonSerializers)
+                {
+                    if (reader.HasAnnotation(valuePair.Key))
+                    {
```
Comment (HomayounA) [inline Amazon.Ion.ObjectMapper/IonSerializer.cs:270]: Is there a method to get the current annotation on reader and looking up that in the options.AnnotatedIonSerializers, rather than looping on it?

File: Amazon.Ion.ObjectMapper/IonSerializer.cs (hunk 1 above)
Comment (simonz-bq) [reply Amazon.Ion.ObjectMapper/IonSerializer.cs]: Added as a part of the rework.

File: Amazon.Ion.ObjectMapper.Test/IonSerializerTest.cs (hunk 2 above)
Comment (simonz-bq) [reply Amazon.Ion.ObjectMapper.Test/IonSerializerTest.cs:60]: Done.

File: Amazon.Ion.ObjectMapper/IonSerializer.cs (hunk 3 above)
Comment (simonz-bq) [reply Amazon.Ion.ObjectMapper/IonSerializer.cs:270]: You have to loop on either the list of annotations or the list of annotated ion serializers.

In this case the list of annotated ion serializers is always the relevant information and the annotations aren't, so I decided to loop on that.

**Comment by simonz-bq** [comment]
Converting this to a draft for now until #22 is completed. Rebasing is not easily possible.
